Cython/Compiler/ExprNodes.py

```diff
@@ -13619,6 +13657,59 @@ def generate_operation_code(self, code, result_code,
             else:
                 code.putln(statement)
 
+    def generate_struct_code(self, operand1_name, operand2_name, operand1_type, operand2_type, op, code):
+        return '(' + (' && ' if op == '==' else ' || ').join([self.generate_ctype_code(operand1_name + '.' + member.name, operand2_name + '.' + member.name, member.type, member.type, op, code) for member in operand2_type.scope.var_entries]) + ')'
+
+    def generate_ctuple_code(self, operand1_name, operand2_name, operand1_type, operand2_type, op, code):
+        if not operand1_type.assignable_from(operand2_type) and not (op in ('in', 'not_in') and any([operand1_type.assignable_from(type_) for type_ in operand2_type.components])):
+            error(self.pos, "cannot compare ctuples of incompatible types")
+
+        temps = list()
+
+        nodes = dict()
+        for itr in range(len(operand2_type.components)):
+            if op in ('in', 'not_in') and not operand1_type.assignable_from(operand2_type.components[itr]):
+                continue
+
+            node = IndexNode(self.pos, base=RawCNameExprNode(self.pos, cname=operand2_name, type=operand2_type), index=IntNode(self.pos, value=str(itr), constant_result=itr, type=PyrexTypes.c_py_ssize_t_type))
+            node.analyse_types(code.funcstate.scope)
```

Comment:
Again, this is way too late in the translation process. This should be done much earlier, either during type analysis (usually) or in a later optimisation phase (if that's what it is).

Reply:
Done.